src/unknown.rs

```diff
@@ -0,0 +1,149 @@
+use alloc::vec::Vec;
+
+use core::fmt::Debug;
+use core::usize;
+
+use bytes::{Buf, BufMut};
+
+use crate::encoding::{decode_varint, encode_key, encode_varint, key_len, WireType};
+use crate::DecodeError;
+
+pub struct UnknownField {
+    tag: u32,
+    wire_type: WireType,
+    bytes: Vec<u8>,
+}
+
+impl Debug for UnknownField {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("UnknownField")
+            .field("tag", &self.tag)
+            .field("wire_type", &self.wire_type)
+            .field("bytes", &self.bytes)
+            .finish()
+    }
+}
+
+impl Clone for UnknownField {
+    fn clone(&self) -> Self {
+        Self {
+            tag: self.tag.clone(),
+            wire_type: self.wire_type.clone(),
+            bytes: self.bytes.clone(),
+        }
+    }
+}
+
+impl PartialEq for UnknownField {
+    fn eq(&self, other: &Self) -> bool {
+        self.tag == other.tag && self.wire_type == other.wire_type && self.bytes == other.bytes
+    }
+}
+
+impl UnknownField {
+    fn encoded_len(&self) -> usize {
+        key_len(self.tag) + self.bytes.len()
+    }
+}
+
+pub struct UnknownFields {
+    fields: Vec<UnknownField>,
+}
+
+impl Debug for UnknownFields {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("UnknownFields")
+            .field("fields", &self.fields)
+            .finish()
+    }
+}
+
+impl Clone for UnknownFields {
+    fn clone(&self) -> Self {
+        Self {
+            fields: self.fields.clone(),
+        }
+    }
+}
+
+impl PartialEq for UnknownFields {
+    fn eq(&self, other: &Self) -> bool {
+        self.fields == other.fields
```

Comment:
This assumes that we treat the unknown fields as different if their order differs.